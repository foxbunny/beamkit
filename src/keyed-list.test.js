const h = require('./h');
const keyedList = require('./keyed-list');
const { prettyPrint } = require('html');

const KEEP = 'keep';
const MOVE_BEFORE = 'move before';
const MOVE_AFTER = 'move after';
const REMOVE = 'remove';
const CREATE = 'create';
const APPEND = 'append';

const CASES = [
  [
    'identical lists',
    ['A', 'B', 'C', 'D', 'E'],
    ['A', 'B', 'C', 'D', 'E'],
    [
      [KEEP, 'A'],
      [KEEP, 'B'],
      [KEEP, 'C'],
      [KEEP, 'D'],
      [KEEP, 'E'],
    ],
  ],
  [
    'completely empty old',
    [],
    ['A', 'B', 'C', 'D', 'E'],
    [
      [APPEND, 'A'],
      [APPEND, 'B'],
      [APPEND, 'C'],
      [APPEND, 'D'],
      [APPEND, 'E'],
    ],
  ],
  [
    'completely empty target',
    ['A', 'B', 'C', 'D', 'E'],
    [],
    [
      [REMOVE, 'A'],
      [REMOVE, 'B'],
      [REMOVE, 'C'],
      [REMOVE, 'D'],
      [REMOVE, 'E'],
    ]
  ],
  [
    'complete swap',
    ['A', 'B', 'C', 'D', 'E'],
    ['F', 'G', 'H', 'I', 'J'],
    [
      [CREATE, 'F', 'A'],
      [CREATE, 'G', 'A'],
      [CREATE, 'H', 'A'],
      [CREATE, 'I', 'A'],
      [CREATE, 'J', 'A'],
      [REMOVE, 'A'],
      [REMOVE, 'B'],
      [REMOVE, 'C'],
      [REMOVE, 'D'],
      [REMOVE, 'E'],
    ]
  ],
  [
    'swap middle',
    ['A', 'B', 'C', 'D', 'E'],
    ['A', 'B', 'D', 'C', 'E'],
    [
      [KEEP, 'A'],
      [KEEP, 'B'],
      [KEEP, 'E'],
      [MOVE_AFTER, 'C', 'D'],
      [KEEP, 'D'],
    ],
  ],
  [
    'middle subset',
    ['A', 'B', 'C', 'D', 'E'],
    ['B', 'C', 'D'],
    [
      [MOVE_BEFORE, 'B', 'A'],
      [MOVE_BEFORE, 'C', 'A'],
      [MOVE_BEFORE, 'D', 'A'],
      [REMOVE, 'A'],
      [REMOVE, 'E'],
    ],
  ],
  [
    'single item overlap',
    ['A', 'B', 'C'],
    ['C', 'D', 'E'],
    [
      [MOVE_BEFORE, 'C', 'A'],
      [CREATE, 'D', 'A'],
      [CREATE, 'E', 'A'],
      [REMOVE, 'A'],
      [REMOVE, 'B'],
    ],
  ],
  [
    'remove initial and from middle',
    ['A', 'B', 'C', 'D', 'E'],
    ['B', 'C', 'E'],
    [
      [KEEP, 'E'],
      [MOVE_BEFORE, 'B', 'A'],
      [MOVE_BEFORE, 'C', 'A'],
      [REMOVE, 'A'],
      [REMOVE, 'D'],
    ],
  ],
  [
    'remove initial',
    ['A', 'B', 'C', 'D', 'E'],
    ['B', 'C', 'D', 'E'],
    [
      [KEEP, 'E'],
      [KEEP, 'D'],
      [KEEP, 'C'],
      [KEEP, 'B'],
      [REMOVE, 'A'],
    ],
  ],
  [
    'append',
    ['A', 'B', 'C', 'D', 'E'],
    ['A', 'B', 'C', 'D', 'E', 'F'],
    [
      [KEEP, 'A'],
      [KEEP, 'B'],
      [KEEP, 'C'],
      [KEEP, 'D'],
      [KEEP, 'E'],
      [APPEND, 'F'],
    ],
  ],
];

describe('keyedList', () => {
  describe('editList', () => {
    const accumulator = () => {
      const changes = [];

      return {
        keep(key) {
          changes.push([KEEP, key]);
        },
        moveBefore(key, pos) {
          changes.push([MOVE_BEFORE, key, pos]);
        },
        moveAfter(key, pos) {
          changes.push([MOVE_AFTER, key, pos]);
        },
        remove(key) {
          changes.push([REMOVE, key]);
        },
        create(key, pos) {
          changes.push([CREATE, key, pos]);
        },
        append(key) {
          changes.push([APPEND, key]);
        },
        getChanges() {
          return changes;
        },
      };
    };

    test.each(CASES)(
      '%s',
      (_title, oldKeys, newKeys, editList) => {
        const acc = accumulator();
        keyedList.editList(
          oldKeys.slice(),
          newKeys.slice(),
          acc,
        );

        expect(acc.getChanges()).toEqual(editList);
      }
    );

    test('append', () => {
      const acc = accumulator();
      keyedList.editList(
        ['A', 'B', 'C', 'D', 'E'],
        ['A', 'B', 'C', 'D', 'E', 'F'],
        acc
      );

      expect(acc.getChanges()).toEqual([
        [KEEP, 'A'],
        [KEEP, 'B'],
        [KEEP, 'C'],
        [KEEP, 'D'],
        [KEEP, 'E'],
        [APPEND, 'F'],
      ]);
    });
  });

  describe('update', () => {
    const renderItem = item => (
      <li>{item.name}</li>
    );

    const List = ({ data }) => (
      <ul>
        {data.map(renderItem)}
      </ul>
    );

    const update = (node, data, oldData)=> {
      if (data.name === oldData.name) return;
      node.textContent = data.name;
    };

    const toItem = key => ({ name: key });

    const toKey = item => item.name;

    test.each(CASES)(
      '%s',
      (title, oldKeys, newKeys) => {
        const oldData = oldKeys.map(toItem);
        const newData = newKeys.map(toItem);
        const root = <List data={oldData} />;

        keyedList.update(oldData, newData, root, renderItem, update, toKey);

        expect(prettyPrint(root.outerHTML)).toMatchSnapshot();

        const resultKeys = [];
        root.childNodes.forEach(node => {
          resultKeys.push(node.textContent);
        });
        expect(resultKeys).toEqual(newKeys);
      }
    );

    test('use key property', () => {
      const oldData = ['A', 'B', 'C', 'D', 'E'].map(item => ({
        key: item,
        name: item,
      }));
      const newData = ['A', 'B', 'C', 'D', 'E', 'F'].map(item => ({
        key: item,
        name: item,
      }));

      const root = <List data={oldData} />;

      keyedList.update(oldData, newData, root, renderItem, update);

      const resultKeys = [];
      root.childNodes.forEach(node => {
        resultKeys.push(node.textContent);
      });
      expect(resultKeys).toEqual(['A', 'B', 'C', 'D', 'E', 'F']);
    });
  });
});
