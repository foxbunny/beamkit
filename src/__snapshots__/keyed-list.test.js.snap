// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyedList update append 1`] = `
"<ul>
    <li>A</li>
    <li>B</li>
    <li>C</li>
    <li>D</li>
    <li>E</li>
    <li>F</li>
</ul>"
`;

exports[`keyedList update complete swap 1`] = `
"<ul>
    <li>F</li>
    <li>G</li>
    <li>H</li>
    <li>I</li>
    <li>J</li>
</ul>"
`;

exports[`keyedList update completely empty old 1`] = `
"<ul>
    <li>A</li>
    <li>B</li>
    <li>C</li>
    <li>D</li>
    <li>E</li>
</ul>"
`;

exports[`keyedList update completely empty target 1`] = `"<ul></ul>"`;

exports[`keyedList update identical lists 1`] = `
"<ul>
    <li>A</li>
    <li>B</li>
    <li>C</li>
    <li>D</li>
    <li>E</li>
</ul>"
`;

exports[`keyedList update middle subset 1`] = `
"<ul>
    <li>B</li>
    <li>C</li>
    <li>D</li>
</ul>"
`;

exports[`keyedList update remove initial 1`] = `
"<ul>
    <li>B</li>
    <li>C</li>
    <li>D</li>
    <li>E</li>
</ul>"
`;

exports[`keyedList update remove initial and from middle 1`] = `
"<ul>
    <li>B</li>
    <li>C</li>
    <li>E</li>
</ul>"
`;

exports[`keyedList update single item overlap 1`] = `
"<ul>
    <li>C</li>
    <li>D</li>
    <li>E</li>
</ul>"
`;

exports[`keyedList update swap middle 1`] = `
"<ul>
    <li>A</li>
    <li>B</li>
    <li>D</li>
    <li>C</li>
    <li>E</li>
</ul>"
`;
